crates/sui-core/src/authority_active/execution_driver/mod.rs

```diff
@@ -1,46 +1,49 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use std::{collections::HashSet, sync::Arc};
-use sui_types::{base_types::TransactionDigest, error::SuiResult, messages::VerifiedCertificate};
-use tracing::{debug, error, info};
-
-use crate::authority::{AuthorityState, PendingDigest};
-use crate::authority_client::AuthorityAPI;
-
-use futures::{stream, StreamExt};
-
 use super::ActiveAuthority;
-
-use tap::TapFallible;
+use crate::authority::AuthorityState;
+use crate::authority_client::AuthorityAPI;
+use async_trait::async_trait;
+use std::{sync::Arc, time::Duration};
+use sui_metrics::spawn_monitored_task;
+use sui_types::{
+    error::{SuiError, SuiResult},
+    messages::VerifiedCertificate,
+};
+use tokio::{
+    sync::Semaphore,
+    time::{sleep, timeout},
+};
+use tracing::{debug, error, info, warn};
 
 #[cfg(test)]
 pub(crate) mod tests;
 
+// Maximum number of certificates to be executed concurrently.
+// TODO: make this configurable.
+const MAX_EXECUTION_CONCURRENCY: usize = 20;
```

Comment:
Let's use num_cpus (https://docs.rs/num_cpus/latest/num_cpus/) as the default instead of hard coding a value.